### PR TITLE
Bug30381

### PR DESCRIPTION
**Pull request by asn-d6**

### CODE REVIEW

File: src/feature/control/control_hs.c [hunk 1]
```diff
@@ -159,3 +159,53 @@ handle_control_onion_client_auth_add(control_connection_t *conn,
...
+  if (BUG(removal_status == REMOVAL_BAD_ADDRESS)) {
+    /* It's a bug because the service addr has already been validated above */
+    control_printf_endreply(conn, 512, "Invalid v3 address \"%s\"",hsaddress);
+  } else if (removal_status == REMOVAL_SUCCESS_NOT_FOUND) {
+    control_printf_endreply(conn, 251, "No credentials for \"%s\"",hsaddress);
+  } else if (removal_status == REMOVAL_SUCCESS) {
+    control_printf_endreply(conn, 250, "OK");
+  } else {
+    tor_assert_nonfatal_unreached();
+  }
```
Comment (dgoulet-tor) [inline src/feature/control/control_hs.c]: Same as before about the `switch()`

File: src/feature/hs/hs_client.h [hunk 2]
```diff
@@ -35,8 +35,12 @@ typedef enum {
...
   /* We successfuly registered these credentials */
   REGISTER_SUCCESS,
-  /* We failed to register these credentials, because they already exist. */
-  REGISTER_FAIL_ALREADY_EXISTS,
+  /* We successfully registered these credentials, but had to replace some
+   * existing ones. */
+  REGISTER_SUCCESS_ALREADY_EXISTS,
+  /* We successfuly registered these credentials, and also decrypted a cached
+   * descriptor. */
+  REGISTER_SUCCESS_ALSO_DECRYPTED,
```
Comment (dgoulet-tor) [inline src/feature/hs/hs_client.h]: `REGISTER_SUCCESS_AND_DECRYPTED` ?

File: src/feature/control/control_hs.c [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  if (strcasecmp(key_type, "x25519")) {
+    control_printf_endreply(conn, 552,
+                            "Unrecognized key type \"%s\"", key_type);
+    goto err;
+  }
+
+  if (base64_decode((char*)privkey->secret_key, sizeof(privkey->secret_key),
+                    key_blob,
+                   strlen(key_blob)) != sizeof(privkey->secret_key)) {
+    control_printf_endreply(conn, 512, "Failed to decode ED25519-V3 key");
```
[the remark this answers is not in the document]
Comment (asn-d6) [reply src/feature/control/control_hs.c]: Fixed in ebeaaffc26.

File: src/feature/control/control_hs.c [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  const char *client_privkey = smartlist_get(args->args, 1);
+  if (parse_private_key_from_control_port(client_privkey,
+                                          &creds->enc_seckey, conn) < 0) {
+    goto err;
+  }
+
+  /* Now let's parse the remaining arguments (variable size) */
+  for (const config_line_t *line = args->kwargs; line; line = line->next) {
+    if (!strcasecmp(line->key, "ClientName")) {
+      /* XXX apply length restriction? */
```
[the remark this answers is not in the document]
Comment (asn-d6) [reply src/feature/control/control_hs.c]: Fixed in ebeaaffc26.

File: src/feature/control/control_hs.c [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+  if (BUG(register_status == REGISTER_FAIL_BAD_ADDRESS)) {
+    /* It's a bug because the service addr has already been validated above */
+    control_printf_endreply(conn, 512, "Invalid v3 address \"%s\"", hsaddress);
+  } else if (register_status == REGISTER_FAIL_ALREADY_EXISTS) {
+    control_printf_endreply(conn, 551, "Client already exists");
+  } else if (register_status == REGISTER_SUCCESS) {
+    control_printf_endreply(conn, 250, "OK");
+  } else {
+    tor_assert_nonfatal_unreached();
+  }
```
[the remark this answers is not in the document]
Comment (asn-d6) [reply src/feature/control/control_hs.c]: Fixed in ebeaaffc26.

File: src/feature/control/control_hs.c (hunk 1 above)
Comment (asn-d6) [reply src/feature/control/control_hs.c]: Fixed in ebeaaffc26.

File: src/feature/hs/hs_client.h (hunk 2 above)
Comment (asn-d6) [reply src/feature/hs/hs_client.h]: Fixed in 26e45d053e.

**Pull request closed without merging**